boa/src/builtins/array/mod.rs

```diff
@@ -495,11 +501,8 @@ impl Array {
         // 3. Let n be 0.
         let mut n = 0;
         // 4. Prepend O to items.
-        let mut items = args
-            .iter()
-            .flat_map(|item_val| item_val.to_object(context))
-            .collect::<VecDeque<GcObject>>();
-        items.push_front(obj);
+        let mut items = vec![Value::from(this)];
+        items.extend_from_slice(args);
         // 5. For each element E of items, do
         for item in items {
```

Comment:
Why do we add `this` to the items, the this value is separate from items argument, this is just `args: &[Value]`

Reply:
It says to prepend this to items?

Reply:
In step 4.

Reply:
> It says to prepend this to items?
> `4. Prepend O to items.`, hence why I figured items must be an array of objects.

Ah, sorry missed that. But we should append `obj` not `this` since it is the result of `.to_object()`.

Also we should be able to do something like:
```rust
for item in [obj.into()].chain(args) {
```
so we do not allocate an vec.

Reply:
@HalidOdat Unless I'm misunderstanding something, this is how we'd expect concat to work, where it would join the left and right?

Reply:
> @HalidOdat Unless I'm misunderstanding something, this is how we'd expect concat to work, where it would join the left and right?

Could you explain what you mean? `[obj.into()].chain(args)` this just joins to rust iterators together.

Reply:
> > It says to prepend this to items?
> > `4. Prepend O to items.`, hence why I figured items must be an array of objects.
> 
> Ah, sorry missed that. But we should append `obj` not `this` since it is the result of `.to_object()`.

But this seems kind of backwards, we initially convert this, from a Value to an Object, and then convert it back to a value for some reason? Instead of just using the Value from before?

Reply:
> But this seems kind of backwards, we initially convert this, from a Value to an Object, and then convert it back to a value for some reason? Instead of just using the Value from before?

The `this` could be an object but it could also we a string for example depeands on how we call it (`Array.prototype.call("I am a string")`), but `obj` will always we a object since it is returned from `to_object()`